core/ecschema-editing/src/Differencing/SchemaDifference.ts

```diff
@@ -74,26 +76,41 @@ export namespace SchemaDifference {
    * @internal
    */
   export async function fromSchemaChanges(targetSchema: Schema, schemaChanges: SchemaChanges): Promise<SchemaDifferences> {
-    const schemaDifference: SchemaDifferences = {
-      sourceSchemaName: schemaChanges.schema.schemaKey.toString(),
-      targetSchemaName: targetSchema.schemaKey.toString(),
-      changes: [],
-      conflicts: [],
-    };
+    const changes: AnySchemaDifference[] = [];
+    const conflicts: SchemaDifferenceConflict[] = [];
 
-    const visitor = new SchemaDiagnosticVisitor(schemaDifference);
+    const visitor = new SchemaDiagnosticVisitor(changes, conflicts);
     for(const diagnostic of schemaChanges.allDiagnostics) {
       visitor.visit(diagnostic);
     }
 
-    return schemaDifference;
+    return {
+      sourceSchemaName: schemaChanges.schema.schemaKey.toString(),
+      targetSchemaName: targetSchema.schemaKey.toString(),
+      changes: changes.length > 0 ? changes : undefined,
+      conflicts: conflicts.length > 0 ? conflicts : undefined,
+    };
   }
 }
 
-type MutualPartial<T> = {
-  -readonly [P in keyof T]?: T[P];
+type Editable<T> = {
+  -readonly [P in keyof T]: T[P];
+};
+
+type ExtractTypeName<T extends SchemaItemType> =
+  { [K in keyof typeof SchemaItemType as typeof SchemaItemType[K]]: K }[T];
+
+type SchemaItemProperties<T extends SchemaItemProps> = {
+  [P in keyof Editable<Omit<T, keyof Omit<SchemaItemProps, "label" | "description" | "customAttributes">>>]: T[P]
 };
 
+type GroupByDifferenceType<T extends  { changeType: DifferenceType, difference: unknown }> =
+  T extends { changeType: infer R extends DifferenceType }
+    ? R extends "modify"
+      ? Omit<T, "difference"> & { difference: Partial<T["difference"]> }
+      : T
+    : never;
+
```

Comment:
I don't think the combination of all of this is going to be maintainable.  Having the 'Any' types were always designed to be discriminated unions: https://basarat.gitbook.io/typescript/type-system/discriminated-unions. This just makes the whole thing way complex